Return only isolated cpus in podresources interface

AlexeyPerevalov opened the pull request:
/kind bug

This change clarifies how which cpu ids are returned by podresourcess interface.
This clarification and change is necessary to avoid requesting pod spec in external daemon to check QoS of the pod, since before this change we were able to distinguish cpus from shared pools and exclusively isolated pool only by QoS Guaranteed.

- Issue: https://github.com/kubernetes/kubernetes/issues/102190

```release-note
Podresources interface was changed, now it returns only isolated cpus
```

Commit:
Return only isolated cpus in podresources interface

Co-Authored-by: Swati Sehgal <swsehgal@redhat.com>
Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

kikisdeliveryservice commented:
So this is related to KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2043-pod-resource-concrete-assigments

Tracked by: https://github.com/kubernetes/enhancements/issues/2043

kikisdeliveryservice commented:
Is there a reason this commit isn't included in: https://github.com/kubernetes/kubernetes/pull/95734 which touches the same file?

AlexeyPerevalov commented:
> 
> 
> Is there a reason this commit isn't included in: #95734 which touches the same file for the same enhancement?
I submitted a new KEP modification https://github.com/kubernetes/enhancements/pull/2201, special for this change, since PR #95734 is already a big enough. Initially we decided to split implementation of the KEP 2043 into several PR to simplify merge each PR.

kikisdeliveryservice commented:
/triage accepted

AlexeyPerevalov commented:
/test pull-kubernetes-bazel-test

AlexeyPerevalov commented:
/test pull-kubernetes-e2e-kind

AlexeyPerevalov commented:
kind/bug

AlexeyPerevalov commented:
/remove-lifecycle stale

swatisehgal commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -462,5 +462,9 @@ func (m *manager) updateContainerCPUSet(containerID string, cpus cpuset.CPUSet)
 
 func (m *manager) GetCPUs(podUID, containerName string) cpuset.CPUSet {
-	return m.state.GetCPUSetOrDefault(podUID, containerName)
+	if result, ok := m.state.GetCPUSet(string(podUID), containerName); ok {
```
We should add podresource API end to end tests to explain and emphasize the rationale behind this change.

fromanirh replied:
Codewise I see no issue with this change. The question is: what is a "concrete resource assignment" for CPUs? this is what concerns podresources API. I never found a very strong definition, so the best shot is trying to understand the original intent of the API, [which was meant to track devices](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/606-compute-device-assignment#user-stories).
So, what is a "cpu assigment" for a container?
In the current form, the API is telling its consumers on which CPUs a container is allowed to run. This tells us little, because we can't distinguish easily exclusive allocated cpus or not.
With this change, we report only the exclusively allocated cpus, which is
1. closer to the original API semantics as I understand it and
2.  gives MORE information to the caller:
    a. if a container has exclusively allocated CPUs, well, it can run only on those cpus. All the information is available.
   b. if a container has NOT exclusively allocated CPUs, it can run on all the allocatable-non-exclusively-allocated CPUs, but this set can constructed using the information provided by the podresources API: get allocatable cpus from `GetAllocatable`, then remove all the exclusive sets we return now. This makes the API self-sufficient and in a better shape than it is today.

For all these reasons, I think this is a good change and I'm fine with this direction.

swatisehgal commented:
I have put some more thought into this and think that it is not appropriate to make change to the behavior of GetCPUs as for clients and sys admins information related CPUs that non-guaranteed pods/ guaranteed pods in shared pool are running on might be useful for certain use cases. CPU Information like this is exposed in the CPU Manager state file and also on the pod cgroup so it would be hard to predict how this information might be used. 

Based on that it appears the best way to address https://github.com/kubernetes/kubernetes/issues/102190 is to introduce another endpoint like GetFreeCPUs() or another boolean field in the `ContainerResources` to be able to provide a way to determine if the container has exclusive CPUs or CPUs belong to the shared pool.

fromanirh commented:
> I have put some more thought into this and think that it is not appropriate to make change to the behavior of GetCPUs as for clients and sys admins information related CPUs that non-guaranteed pods/ guaranteed pods in shared pool are running on might be useful for certain use cases. CPU Information like this is exposed in the CPU Manager state file and also on the pod cgroup so it would be hard to predict how this information might be used.
> 
> Also had an offline discussion with @AlexeyPerevalov on this and we agree that the best way to address #102190 is to introduce another endpoint like GetFreeCPUs() or another boolean field in the `ContainerResources` to be able to provide a way to determine if the container has exclusive CPUs or CPUs belong to the shared pool.

This is surely the safest and most conservative approach, so I think it should be fine :)
I was wondering if we can't just add a flag in `ListPodResourcesRequest` (https://github.com/kubernetes/kubernetes/blob/release-1.21/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto#L33) and `AllocatableResourcesRequest` (https://github.com/kubernetes/kubernetes/blob/release-1.21/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto#L24) it should be backward compatible and in the boundaries of what we can change in the v1 API.

swatisehgal commented:
> I was wondering if we can't just add a flag in `ListPodResourcesRequest` (https://github.com/kubernetes/kubernetes/blob/release-1.21/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto#L33) and `AllocatableResourcesRequest` (https://github.com/kubernetes/kubernetes/blob/release-1.21/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto#L24) it should be backward compatible and in the boundaries of what we can change in the v1 API.

My thought process was to add a flag at container level as when the client acquires information about the CPUs allocated to a container they would want the ability to check if the CPUs belong to shared pool or are exclusive CPUs. 
```
// ContainerResources contains information about the resources assigned to a container
message ContainerResources {
    string name = 1;   
    repeated ContainerDevices devices = 2;
    repeated int64 cpu_ids = 3;
    // new field
    bool exclusiveCPUs = 4;
}
```

or 
```
// new endpoint
rpc GetSharedCPUPool() returns (SharedCPUResponse) {}
message SharedCPUResponse {
    repeated int64 shared_cpu_ids = 1;
}
```

fromanirh commented:
> > I was wondering if we can't just add a flag in `ListPodResourcesRequest` (https://github.com/kubernetes/kubernetes/blob/release-1.21/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto#L33) and `AllocatableResourcesRequest` (https://github.com/kubernetes/kubernetes/blob/release-1.21/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto#L24) it should be backward compatible and in the boundaries of what we can change in the v1 API.
> 
> My thought process was to add a flag at container level as when the client acquires information about the CPUs allocated to a container they would want the ability to check if the CPUs belong to shared pool or are exclusive CPUs. Something like below:
> 
> ```
> // ContainerResources contains information about the resources assigned to a container
> message ContainerResources {
>     string name = 1;   
>     repeated ContainerDevices devices = 2;
>     repeated int64 cpu_ids = 3;
>     // new field
>     bool exclusiveCPUs = 4;
> }
> ```
> 
> or a new endpoint like below:
> 
> ```
> // new endpoint
> rpc GetFreeCPUs() [or GetSharedCPUPool()] returns (SharedCPUResponse) {}
> message SharedCPUResponse {
>     repeated int64 shared_cpu_ids = 1;
> }
> ```

IMHO, from the perspective of maximizing backward compatibility and minimizing the changes, I think the first option (exclusiveCPUs flag) is better.
With respect to adding the flag to the request objects (https://github.com/kubernetes/kubernetes/pull/97415#issuecomment-847873651) this change give back more information, so it seems a bit better.

fromanirh commented:
> > I have put some more thought into this and think that it is not appropriate to make change to the behavior of GetCPUs as for clients and sys admins information related CPUs that non-guaranteed pods/ guaranteed pods in shared pool are running on might be useful for certain use cases. CPU Information like this is exposed in the CPU Manager state file and also on the pod cgroup so it would be hard to predict how this information might be used.
> > Also had an offline discussion with @AlexeyPerevalov on this and we agree that the best way to address #102190 is to introduce another endpoint like GetFreeCPUs() or another boolean field in the `ContainerResources` to be able to provide a way to determine if the container has exclusive CPUs or CPUs belong to the shared pool.
> 
> This is surely the safest and most conservative approach, so I think it should be fine :)
> I was wondering if we can't just add a flag in `ListPodResourcesRequest` (https://github.com/kubernetes/kubernetes/blob/release-1.21/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto#L33) and `AllocatableResourcesRequest` (https://github.com/kubernetes/kubernetes/blob/release-1.21/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto#L24) it should be backward compatible and in the boundaries of what we can change in the v1 API.

Thinking about it, extra flags to `AllocatableResourcesRequest` are probably not needed.

swatisehgal commented:
> Thinking about it, extra flags to `AllocatableResourcesRequest` are probably not needed.

Thanks for your input on this! I think I understand your idea a bit better now but if we introduce a new boolean flag like `reportExclusiveCPUs` in the `ListPodResourcesRequest` in order to get the full view of pods that have shared CPUs and the ones that have exclusive CPUs we would need two calls which might be a shortcoming.

fromanirh commented:
> > Thinking about it, extra flags to `AllocatableResourcesRequest` are probably not needed.
> 
> Thanks for your input on this! I think I understand your idea a bit better now but if we introduce a new boolean flag like `reportExclusiveCPUs` in the `ListPodResourcesRequest` in order to get the full view of pods that have shared CPUs and the ones that have exclusive CPUs we would need two calls which might be a shortcoming.

Yes. Pretty much. This is why I believe the approach you suggested with the new extra `bool` flag is better.

kad commented:
The meaning of "exclusive" or "shared" is very specific to CPU Manager internal state (policy). Imagine that you might have another policy that has multiple shared pools of CPUs, or groupping several workloads into subset pool of CPUs (exclusive core allocation for several burstable Pods).... all those assumptions will be broken.

If someone wants to re-construct internal state of CPU Manager out of PodResources API, it can be done by boolean operations on assigned/total CPU sets already present in PodResources.

swatisehgal commented:
> The meaning of "exclusive" or "shared" is very specific to CPU Manager internal state (policy). Imagine that you might have another policy that has multiple shared pools of CPUs, or groupping several workloads into subset pool of CPUs (exclusive core allocation for several burstable Pods).... all those assumptions will be broken.

This is a valid point! So far I have been proposing the changes based on the current implementation of CPU Manager and the fact that it tackles guaranteed pods and they could belong to two pools "exclusive" and "shared" but we should definitely ensure that the API is robust enough to tackle more policies which could introduce more pools. 
I think we can assume that there is a unique string identifier ( cpu_pool_id) that represents each pool. Do you think that is reasonable? Based on the current CPU Manager implementation cpu_pool_id could be "shared" and "exclusive" and we could have the API like below:
```
message ContainerResources {
    string name = 1;   
    repeated ContainerDevices devices = 2;
    repeated int64 cpu_ids = 3;
    // new field representing the CPU pool
    string cpu_pool_id = 4;
}
```
or if a new endpoint is like below:

```
// new endpoint
rpc GetCPUPools() returns (CPUPoolResponse) {}
message CPUPoolResponse {
    // map where key is cpu_pool_id and value is the cpuids that form that pool
    map<string, repeated int64> cpu_pools = 2; 
}
```

swatisehgal commented:
> If someone wants to re-construct internal state of CPU Manager out of PodResources API, it can be done by boolean operations on assigned/total CPU sets already present in PodResources.


Well the goal here to be able to determine the free CPUs on a system that would not be possible without knowing exactly which pool the pod belonged to. Let me explain this with an example, consider a scenario:

A cluster with a node with 8 CPUs (CPUID=0 configured as reserved) with pods running as follows:
pod-1 A guaranteed pod requesting 500m CPU // allocated CPUs from the shared pool
pod-2 A guaranteed pod requesting 2 CPU        // allocated exclusive CPUs
pod-3 A guaranteed pod requesting 2 CPU        // allocated exclusive CPUs

After querying the pod resource API we see the following:
```
GetAllocatableResources() returns [1 2 3 4 5 6 7]
List Returns: 
[{pod-1 rte [{container-1 [{cpu [7 0 3 6]}]}]} 
{pod-2 rte [{container-1 [{cpu [4 5]}]}]} 
{pod-3 rte [{container-1 [{cpu [1 2]}]}]}]
```

How do we determine just from the pod resource API if pod-1 belongs to the shared pool and not exclusive pool because  from podresource API it looks exactly the same as a guaranteed pod that requested 4 CPUs would look like.
Without knowing the pool it belonged to, it is not possible to correctly evaluate the free CPUs on a system unless I am missing something?

AlexeyPerevalov commented:
> After querying the pod resource API we see the following:

AlexeyPerevalov commented:
> > If someone wants to re-construct internal state of CPU Manager out of PodResources API, it can be done by boolean operations on assigned/total CPU sets already present in PodResources.
> 
> Well the goal here to be able to determine the free CPUs on a system that would not be possible without knowing exactly which pool the pod belonged to. Let me explain this with an example, consider a scenario:
> 
> A cluster with a node with 8 CPUs (CPUID=0 configured as reserved) with pods running as follows:
> pod-1 A guaranteed pod requesting 500m CPU // allocated CPUs from the shared pool
> pod-2 A guaranteed pod requesting 2 CPU // allocated exclusive CPUs
> pod-3 A guaranteed pod requesting 2 CPU // allocated exclusive CPUs
> 
> After querying the pod resource API we see the following:
> 
> ```
> GetAllocatableResources() returns [1 2 3 4 5 6 7]
> List Returns: 
> [{pod-1 rte [{container-1 [{cpu [7 0 3 6]}]}]} 
> {pod-2 rte [{container-1 [{cpu [4 5]}]}]} 
> {pod-3 rte [{container-1 [{cpu [1 2]}]}]}]
> ```
> 
> How do we determine just from the pod resource API if pod-1 belongs to the shared pool and not exclusive pool because from podresource API it looks exactly the same as a guaranteed pod that requested 4 CPUs would look like.
> Without knowing the pool it belonged to, it is not possible to correctly evaluate the free CPUs on a system unless I am missing something?

Maybe just add cpu_pool_id not only into ContainerResources, but also into AllocatableResourcesResponse, as I remember we decided to not reuse ContainerResources in GetAllocatableResources, but just to copy its content. So here we've to follow.

kad commented:
> ```
> message ContainerResources {
>     string name = 1;   
>     repeated ContainerDevices devices = 2;
>     repeated int64 cpu_ids = 3;
>     // new field representing the CPU pool
>     string cpu_pool_id = 4;
> }
> ```

cpu_pool_id is redundant in this structure. BTW, in policies that support mixed allocations (exclusive + shared) container would belong to multiple "pools".
So, repeated list of cpu_ids, and actually mems_ids for memory regions (once memory manager would expose that) is the thing that should show what is available to the container. The rest can be re-calculated from other data.

> or if a new endpoint is like below:
> 
> ```
> // new endpoint
> rpc GetCPUPools() returns (CPUPoolResponse) {}
> message CPUPoolResponse {
>     // map where key is string values corresponding to cpu_pool_id and value is the cpuids that form that pool
>     map<string, repeated int64> cpu_pools = 2; 
> }
> ```

I think it would be better instead explicitly exposing CPU manager internal structures to improve GetAllocatableResources. E.g. if current policy in CPU Manager treats them as different pools, it could return as `[[1,2],[4,5],[0,3,6,7]]` for example.
That would give enough information to re-calculate split of resources in external customers if it has any value there.

swatisehgal commented:
> Maybe just add cpu_pool_id not only into ContainerResources, but also into AllocatableResourcesResponse, as I remember we decided to not reuse ContainerResources in GetAllocatableResources, but just to copy its content. So here we've to follow.

GetAllocatableResources returns information of all the allocatable resources and the idea is that it rarely changes as the resources exposed to the node changes rarely however depending on the allocation of the pods the pools shrink and grow. I created [this](https://github.com/kubernetes/kubernetes/issues/102191) issue and added documentation fixed to emphasize this behavior. So having it in the ContainerResources is enough because all we need to evaluate is whether the container has exclusive CPUs or not.

swatisehgal commented:
> > ```
> > message ContainerResources {
> >     string name = 1;   
> >     repeated ContainerDevices devices = 2;
> >     repeated int64 cpu_ids = 3;
> >     // new field representing the CPU pool
> >     string cpu_pool_id = 4;
> > }
> > ```
> 
> cpu_pool_id is redundant in this structure. BTW, in policies that support mixed allocations (exclusive + shared) container would belong to multiple "pools".
> So, repeated list of cpu_ids, and actually mems_ids for memory regions (once memory manager would expose that) is the thing that should show what is available to the container. The rest can be re-calculated from other data.

I think we have diverged a bit from the main goal here.  The goal here is to determine whether or not a pod and more specifically a container has exclusive CPUs, which was done in my initial proposal [here](https://github.com/kubernetes/kubernetes/pull/97415#issuecomment-847891218). In order to do the accounting appropriately to determine the free CPUs on a node we need to what CPUs have been exclusively allocated which a boolean flag would allow us to do. `GetAllocatableResources` returns all the CPUs exposed to the node including the allocatedCPUs and the unallocated CPUs.

> I think it would be better instead explicitly exposing CPU manager internal structures to improve GetAllocatableResources. E.g. if current policy in CPU Manager treats them as different pools, it could return as `[[1,2],[4,5],[0,3,6,7]]` for example.
> That would give enough information to re-calculate split of resources in external customers if it has any value there.

Please refer to my comment about `GetAllocatableResources` [here](https://github.com/kubernetes/kubernetes/pull/97415#issuecomment-848757872) and the intent of having GetAllocatableResources. Furthermore even if we split the response into slices of strings, how would you know which pool is which? Which pool represents the allocatedCPUs and which ones represent the availableCPUs/freeCPUs?

kad commented:
> > cpu_pool_id is redundant in this structure. BTW, in policies that support mixed allocations (exclusive + shared) container would belong to multiple "pools".
> > So, repeated list of cpu_ids, and actually mems_ids for memory regions (once memory manager would expose that) is the thing that should show what is available to the container. The rest can be re-calculated from other data.
> 
> I think we have diverged a bit from the main goal here. The goal here is to determine whether or not a pod and more specifically a container has exclusive CPUs, which was done in my initial proposal [here](https://github.com/kubernetes/kubernetes/pull/97415#issuecomment-847891218). In order to do the accounting appropriately to determine the free CPUs on a node we need to what CPUs have been exclusively allocated which a boolean flag would allow us to do. `GetAllocatableResources` returns all the CPUs exposed to the node including the allocatedCPUs and the unallocated CPUs.

The "exclusive" is just a small dynamically allocated pool that is used by single container.
The "shared" pool is also shrinks and grows based on slicing out or returning back the vCPUs from exclusive pools.

To calculate "free" CPUs is possible by iterating via resources assigned to containers and union/intersection operation with Allocatable reported pools. e.g. "exclusive" will match only with single container, "shared" will be matching with multiple containers.

> > I think it would be better instead explicitly exposing CPU manager internal structures to improve GetAllocatableResources. E.g. if current policy in CPU Manager treats them as different pools, it could return as `[[1,2],[4,5],[0,3,6,7]]` for example.
> > That would give enough information to re-calculate split of resources in external customers if it has any value there.
> 
> Please refer to my comment about `GetAllocatableResources` [here](https://github.com/kubernetes/kubernetes/pull/97415#issuecomment-848757872) that explains the intent of having GetAllocatableResources. Furthermore even if we split the response into slices of strings, how would you know which pool is which? Which pool represents the allocatedCPUs and which ones represent the availableCPUs/freeCPUs?

For current policy - any vCPU mentioned in "shared" pool is practically available for allocation. But that might be not the case for any other policy. Even for static policy, even if you oversimplify Allocatable-AllocatedExclusive that would be not the correct number. Think about Burstable QoS containers - you can't squeeze shared pool to less than sum(Burstable_CPU_Requests).

So, if you want to recalculate correctly internal state of CPU Manager static policy - you anyway would need to go over container requests/limits + allowed CPUsets to determine which amount of resources they are really using.

swatisehgal commented:
> The "exclusive" is just a small dynamically allocated pool that is used by single container.
> The "shared" pool is also shrinks and grows based on slicing out or returning back the vCPUs from exclusive pools.

Agree!
> To calculate "free" CPUs is possible by iterating via resources assigned to containers and union/intersection operation with Allocatable reported pools. e.g. "exclusive" will match only with single container, "shared" will be matching with multiple containers.

This is totally right and I understand this but there is a huge "if" here! This relies on the assumption that at a given time there will be at least two pods that obtain CPUs from the shared pool. What if at a given point of time when the client is making this accounting calculation there are all guaranteed pods requesting integral CPUs (and hence allocation CPUs from exclusive pool) and only one pod in the shared pool. This way of calculation would not work then! It would be inappropriate to rely on the cluster pod composition (which can be very dynamic) for evaluating free resources.

> For current policy - any vCPU mentioned in "shared" pool is practically available for allocation. But that might be not the case for any other policy. Even for static policy, even if you oversimplify Allocatable-AllocatedExclusive that would be not the correct number. Think about Burstable QoS containers - you can't squeeze shared pool to less than sum(Burstable_CPU_Requests).

Please note that both CPU Manager and pod resource API at this stage only handles pods that guaranteed pods. Burstable QoS containers are not considered because of exactly the reasons you are highlighting here: the associated complexity that it brings with itself. In order cater to burstable QoS class like you talking about here there would have to be a lot of other things and I am guessing rounds of discussion we would have to consider in CPU Manager in Kubelet even before we start thinking about how it would have to be represented in PodResource API.

> So, if you want to recalculate correctly internal state of CPU Manager static policy - you anyway would need to go over container requests/limits + allowed CPUsets to determine which amount of resources they are really using.

The goal is to enable PodResource API to be able to this. Please refer to the [goals](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2403-pod-resources-allocatable-resources#goals) and the [user stories](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2403-pod-resources-allocatable-resources#user-stories).

ehashman commented:
/assign @swatisehgal 

Agree that this needs tests.

ehashman commented:
As well - the KEP is not approved for this release, the linked KEP PR needs to be merged and set for this release for this work to proceed.

AlexeyPerevalov commented:
Didn't expect, this issue raise so hot discussion. Let me summarize approaches:
1. Do not modify interface signature, but to limit the cpu by only exclusive cpu. This hot fix could be a solution for particular product, but not for upstream.
2. Add boolean field into ContainerResources. Here we found a disadvantage: assumption that pod could belong to only one cpu pool in general is incorrect, we may have a lot of pools, like currently existing external resource managers provide. So from API point of view, this solution too limited.
3. Add string field into ContainerResources or into AllocatableResourcesResponse to represent a set of pool. In this case we have to track several instances of the ContainerResources or AllocatableResourcesResponse per unique pool and devices field in this case looks orphaned. Or we have to move to special structure to represent CPU pool and replace cpu_ids field by field of such new type. It breaks backward compatibility. So it's not a solution too.
4. New endpoint, which describes the pools or some resource domains we may have in the future (GetSharedCPUPool/GetCPUPools), but here I would prefer to not focus on cpu in the name of new endpoint (pool I think is ok), unless we want to introduce new endpoint per every resource type.
Now after sleeping on this issue, I think, the last point is more preferable.

marquiz commented:
> Didn't expect, this issue raise so hot discussion. Let me summarize approaches:
> 
> 1. Do not modify interface signature, but to limit the cpu by only exclusive cpu. This hot fix could be a solution for particular product, but not for upstream.
> 2. Add boolean field into ContainerResources. Here we found a disadvantage: assumption that pod could belong to only one cpu pool in general is incorrect, we may have a lot of pools, like currently existing external resource managers provide. So from API point of view, this solution too limited.
> 3. Add string field into ContainerResources or into AllocatableResourcesResponse to represent a set of pool. In this case we have to track several instances of the ContainerResources or AllocatableResourcesResponse per unique pool and devices field in this case looks orphaned. Or we have to move to special structure to represent CPU pool and replace cpu_ids field by field of such new type. It breaks backward compatibility. So it's not a solution too.
> 4. New endpoint, which describes the pools or some resource domains we may have in the future (GetSharedCPUPool/GetCPUPools), but here I would prefer to not focus on cpu in the name of new endpoint (pool I think is ok), unless we want to introduce new endpoint per every resource type.
> 
> Now after sleeping on this issue, I think, the last point is more preferable.

Hi @AlexeyPerevalov. We found out serious gaps in the current API. Basically, in order to do proper accounting for the resources (and deduce the internal state of affairs kubelet) we need to list container resource requests in addition to just plain resource ids (and why not limits, too) as well as system-reserved in a similar manner. I'm working on a proposal to address this and will pull you (all) in when I have an initial draft ready. In the coming days.

swatisehgal committed:
excludesharedpool: e2e tests: code refactor to handle non-integral CPUs

This patch changes cpuCount to cpuRequest in order to cater to cases
where guaranteed pods make non-integral CPU Requests.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
excludesharedpool: e2e tests: Test cases for pods with non-integral CPUs

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/remove-lifecycle stale

AlexeyPerevalov commented:
/test pull-kubernetes-node-kubelet-serial

AlexeyPerevalov commented:
/retest

AlexeyPerevalov commented:
/test pull-kubernetes-node-kubelet-serial

swatisehgal commented:
Based on the discussion on this PR and on https://github.com/kubernetes/kubernetes/pull/102989 we have decided to avoid any API changes that give too much information about the internal state CPU manager to solve the issue: https://github.com/kubernetes/kubernetes/issues/10219. The implementation in this PR now solves the bug without API changes and has now been updated with e2e tests.

fromanirh commented on test/e2e_node/podresources_test.go:
```diff
@@ -48,7 +48,7 @@ type podDesc struct {
 	resourceName   string
 	resourceAmount int
-	cpuCount       int
+	cpuRequest     float32
```
why? I think we can express the cpuamount in millicores so we can keep the `int`.
This way, say, 0.5 cpus becomes `500` and 2.5 becomes `2500`. If we go this path, let's add a comment documenting the request is in millicores (not whole cpus as it was).

swatisehgal replied:
Sure, that sounds reasonable to me. Will update the e2e tests soon.

fromanirh commented on test/e2e_node/podresources_test.go:
```diff
@@ -739,3 +737,7 @@ func enablePodResourcesFeatureGateInKubelet(f *framework.Framework) (oldCfg *kub
 }
+
+func isIntegral(cpuRequest float32) bool {
+	return cpuRequest == float32(int(cpuRequest))
```
I'm always wary when I see equality check with floats, been burned more times than I'm comfortable recalling.
If we keep the `cpuRequest` as `int`, the check should look like
```
func isIntegral(cpuRequest int) bool {
	return (cpuRequest % 1000) == 0
}
```
which is always safe and easy to verify.

swatisehgal replied:
Definitely!

ffromani reviewed (commented):
This seems a good fix and I fully support it. (full rationale : https://github.com/kubernetes/kubernetes/pull/97415/commits/9b8144b8008a462d415635388266443ba3ee1588#r715463840). I have a minor but still important comment regarding the e2e tests which I'd like to discuss.
Once that is handled I'll be happy to add my `lgtm`.

fromanirh commented:
/lgtm

fromanirh commented:
please add the release note (`NONE` if not needed), otherwise the label "do-not-merge/release-note-label-needed" will prevent the merge

fromanirh commented:
/priority important-longterm

fromanirh commented:
/remove-kind feature
/kind bug

fromanirh commented:
We will need @klueska 's review and approval for this change.

AlexeyPerevalov commented:
/retest

fromanirh commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

fromanirh commented:
/retest

AlexeyPerevalov commented:
/test pull-kubernetes-node-kubelet-serial

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -88,6 +88,10 @@ type Manager interface {
+
+	// GetCPUsetOrDefault returns cpuset which includes cpus from shared pools
+	// as well as exclusively allocated cpus
+	GetCPUsetOrDefault(podUID, containerName string) cpuset.CPUSet
```
Assuming this is true, I feel like better names for these API calls would be:

1.  `GetExclusiveCPUs()`
2.  `GetCPUAffinity()` 

Where `GetExclusiveCPUs()` returns the list of CPUs exclusive to this container, and `GetCPUAffinity()` returns the full list of CPUs this container is able to run on (i.e. the CPUs it is affinitized to). If `GetExclusiveCPUs() == GetCPUAffinity()`, then you know all CPUs assigned to the container are exclusive, otherwise they are not.

However, since we already have `GetCPUs()`as an API call, I am fine repurposing it to mean `GetExclusiveCPUs()` (instead of introducing a whole new call for this purpose). We should definitely update the comment to make this very clear though.

That said, I do think we should rename `GetCPUsetOrDefault()` to `GetCPUAffinity()`. This is much clearer in my opinion, and not as tied to the current implementation / naming of the call in the `state` abstraction.

I'm curious if / how this naming could be extended to devices as well. At present we don't allow sharing of devices, but nothing is preventing us from allowing this in the future. I guess `GetDevices()` could continue to return "exclusive" devices, but a call like `GetDeviceAffinity()` doesn't really make much sense (especially since this call returns a list of many different device types across different vendors). What would be a good name in this circumstance to mean "Shared+Exclusive"?

swatisehgal replied:
> Just to make sure I'm understanding this correctly -- this new call returns what `GetCPUs()` used to return (i.e. the list of CPUs from `m.state.GetCPUSetOrDefault()`).
> 
> And the `GetCPUs()` call has now been changed to only return the list of _exclusive_ CPUs (if there are any).
> 
> Is this correct?

Yes, it's correct!

swatisehgal replied:
> Assuming this is true, I feel like better names for these API calls would be:
> 
> 1. `GetExclusiveCPUs()`
> 2. `GetCPUAffinity()`
> 
> Where `GetExclusiveCPUs()` returns the list of CPUs exclusive to this container, and `GetCPUAffinity()` returns the full list of CPUs this container is able to run on (i.e. the CPUs it is affinitized to). If `GetExclusiveCPUs() == GetCPUAffinity()`, then you know all CPUs assigned to the container are exclusive, otherwise they are not.
> 
> However, since we already have `GetCPUs()`as an API call, I am fine repurposing it to mean `GetExclusiveCPUs()` (instead of introducing a whole new call for this purpose). We should definitely update the comment to make this very clear though.
> That said, I do think we should rename `GetCPUsetOrDefault()` to `GetCPUAffinity()`. This is much clearer in my opinion, and not as tied to the current implementation / naming of the call in the `state` abstraction.

Sure! I will update the names of the API calls as per your suggestion.

swatisehgal replied:
> I'm curious if / how this naming could be extended to devices as well. At present we don't allow sharing of devices, but nothing is preventing us from allowing this in the future. I guess GetDevices() could continue to return "exclusive" devices, but a call like GetDeviceAffinity() doesn't really make much sense (especially since this call returns a list of many different device types across different vendors). What would be a good name in this circumstance to mean "Shared+Exclusive"?

This is interesting and we can prepare ourselves for a case like this in case of devices as well. How about we change the name of GetDevice() to GetExclusiveDevice() and then repurpose GetDevice() API call to correspond to "Shared+Exclusive" devices ?

klueska reviewed (commented):
Hey guys, sorry I'm a little late to the game. In general, this change seems to be going in the right direction, I just have a few questions / comments.

klueska commented:
/assign klueska

fromanirh commented:
Let's look at the serial test failures:
- DynamicKubeletConfiguration: handled by https://github.com/kubernetes/kubernetes/pull/104481 and https://github.com/kubernetes/kubernetes/pull/104153
- Memory manager seems to be a flake, handled and discussed in https://github.com/kubernetes/kubernetes/pull/105193
- SystemNodeCriticalPod deserves more investigation and possibly a fix, I posted https://github.com/kubernetes/kubernetes/pull/103408 some time ago which should help

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -88,6 +88,10 @@ type Manager interface {
 
 	// GetAllocatableCPUs returns the assignable (not allocated) CPUs
 	GetAllocatableCPUs() cpuset.CPUSet
+
+	// GetCPUAffinity returns cpuset which includes cpus from shared pools
+	// as well as exclusively allocated cpus
+	GetCPUAffinity(podUID, containerName string) cpuset.CPUSet
```
Ah, I somehow missed that this was an internal `CPUManager` API. I still think it's a better name, but I somehow thought that it was being exposed all the way through to the podresources API.

And if it's not ... I guess there is now no way for a consumer of the podresources to get the [taskset](https://man7.org/linux/man-pages/man1/taskset.1.html) (i.e. `CPUAffinity`) of a container (unless it has exclusive CPUs and `GetExclusiveCPUs() == GetCPUAffinity()`).

klueska replied:
I think so. I'm a little worried about the fact that we no longer have a way of getting the `taskset` of a container through this API anymore, and I'm a little worried about breaking people that relied on these semantics with the existing call (without giving them a way to get that information at all anymore). What's your sense on this?

swatisehgal replied:
`Taskset` of the container provides information about the CPUs that the container can run on but what it lacking is way to determine whether those CPUs are exclusively allocated to that container or not. This leaves ambiguity in the API and prevents us from meeting the original goals that motivated these changes ([concrete resources](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2403-pod-resources-allocatable-resources#topology-aware-scheduling) and [getAllocatableResources](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2403-pod-resources-allocatable-resources#topology-aware-scheduling) endpoint) in the podresource API in the first place.

My take is that it is appropriate to not consider the shared CPUs when we are talking about concrete resources allocated to a container because the container is not its sole consumer (CPUs are shared with other containers). Also, from client's perspective:

1. If the CPUs are exclusively allocated CPUs we have that information (with the changes proposed in this PR)
2. If the CPUs are not exclusively allocated, CPUs are obtained from the shared pool and this information can be obtained by calling the `GetAllocatableResources` endpoint

So essentially the information is there if the client wants it!

klueska replied:
Yeah, I understand the problem this PR is addressing. I just want to make sure someone can still get the same information they used to from the previous call to `GetGPUs()` before this change (essentially the `taskset` of the container).

To get that same info after this change, they will need to:
1. Call `GetAllocatableResources` to get a list of all the allocatable CPUs
2. Call `GetCPUs` on *all* containers in the system
3. Subtract out all of the CPUs from the `GetCPUs` calls from the `GetAllocatableResources` call

My guess is that most people that even use this API are mostly curious about what exclusive CPUs they have, so maybe this isn't a real issue. I just don't want anyone to be surprised by this change.

swatisehgal replied:
Point taken! I will make sure that this is explicitly captured in the documentation [PR](https://github.com/kubernetes/website/pull/29751) and no one is taken by surprise. I added a [commit](https://github.com/kubernetes/website/pull/29751/commits/daa58fa14ad6d05de1f0038b1f696d1b16286eac) to the documentation PR to take care of this. Does that look like a reasonable way to take care of this?

swatisehgal commented:
/test pull-kubernetes-integration

ffromani reviewed (commented):
/lgtm
LGTM lost because a trivial rebase, restoring
